core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -0,0 +1,315 @@
+        Exception exception = new Exception("Unknown exception in ReadAheadInputStream");
+        try {
+          while (true) {
+            read = underlyingInputStream.read(arr);
+            if (0 != read) break;
+          }
+          handled = true;
+        } catch (Exception ex) {
+          exception = ex;
+        } finally {
+          stateChangeLock.lock();
+          if (read < 0 || (exception instanceof EOFException) ) {
+            endOfStream = true;
+          } else if (!handled) {
+            readAborted = true;
+            readException = exception != null ? exception: new Exception("Unknown exception in ReadAheadInputStream");
+          } else {
+            byteBuffer.limit(read);
+          }
+          readInProgress = false;
+          signalAsyncReadComplete();
+          stateChangeLock.unlock();
+        }
+      }
+    });
+  }
+
+
+  private void signalAsyncReadComplete() {
+    stateChangeLock.lock();
+    try {
+      asyncReadComplete.signalAll();
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  private void waitForAsyncReadComplete() {
+    stateChangeLock.lock();
+    try {
+      if (readInProgress)
+      asyncReadComplete.await();
+    } catch (InterruptedException e) {
+    } finally {
+      stateChangeLock.unlock();
+    }
+  }
+
+  @Override
+  public int read() throws IOException {
+    int val = read(oneByte.get(), 0, 1);
+    if (val == -1) {
+      return -1;
+    }
+    return oneByte.get()[0] & 0xFF;
+  }
+
+  @Override
+  public int read(byte[] b, int offset, int len) throws IOException {
+    if (offset < 0 || len < 0 || offset + len < 0 || offset + len > b.length) {
+      throw new IndexOutOfBoundsException();
+    }
+    if (len == 0) {
+      return 0;
+    }
+    stateChangeLock.lock();
+    try {
+      len = readInternal(b, offset, len);
+    }
+    finally {
+      stateChangeLock.unlock();
+    }
+    return len;
+  }
+
+  /**
+   * Internal read function which should be called only from read() api. The assumption is that
+   * the stateChangeLock is already acquired in the caller before calling this function.
+   */
+  private int readInternal(byte[] b, int offset, int len) throws IOException {
+    assert (stateChangeLock.isLocked());
+    if (!activeBuffer.hasRemaining()) {
+      if (!readInProgress) {
+        // This condition will only be triggered for the first time read is called.
+        readAsync(activeBuffer);
+      }
+      waitForAsyncReadComplete();
+    }
+    if (readAborted) {
+      throw new IOException(readException);
+    }
+    if (isEndOfStream()) {
+      return -1;
+    }
+    len = Math.min(len, activeBuffer.remaining());
+    activeBuffer.get(b, offset, len);
+
+    if (activeBuffer.remaining() <= readAheadThresholdInBytes && !readAheadBuffer.hasRemaining()) {
+      readAsync(readAheadBuffer);
+    }
+    if (!activeBuffer.hasRemaining()) {
+      ByteBuffer temp = activeBuffer;
+      activeBuffer = readAheadBuffer;
```

Comment:
Looks like there is a race condition here: the next `read` will try to read from a buffer that is being read from the underlying input stream.

Reply:
I don't see any race condition here. The readInternal function is protected by the stateChangeLock so only one thread is guaranteed to atomically flip the buffer. Am I missing something?